packages/webapp/src/components/TileDashboard/useDynamicTileVisibility.ts

```diff
@@ -0,0 +1,89 @@
+/*
+ *  Copyright 2024 LiteFarm.org
+ *  This file is part of LiteFarm.
+ *
+ *  LiteFarm is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  LiteFarm is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
+ */
+
+import { useState, useLayoutEffect, RefObject } from 'react';
+import type { TypeCountTile } from './index';
+
+// Source: https://github.com/que-etc/resize-observer-polyfill
+interface ResizeObserver {
+  observe(target: Element): void;
+  unobserve(target: Element): void;
+  disconnect(): void;
+}
+
+interface useDynamicTileVisibilityParams {
+  containerRef: RefObject<HTMLElement>;
+  typeCountTiles: TypeCountTile[];
+  gap?: number;
+  tileWidth?: number; // & more Button width
+  minWidthDesktop?: number;
+  rowsPerView?: {
+    desktop: number;
+    mobile: number;
+  };
+}
+
+export const useDynamicTileVisibility = ({
+  containerRef,
+  typeCountTiles,
+  gap = 4,
+  tileWidth = 90, // & more Button width
+  minWidthDesktop = 600,
+  rowsPerView = { desktop: 1, mobile: 2 },
+}: useDynamicTileVisibilityParams) => {
+  const [visibleIconTiles, setVisibleIconTiles] = useState<TypeCountTile[]>([]);
+  const [hiddenIconTiles, setHiddenIconTiles] = useState<TypeCountTile[]>([]);
+
+  useLayoutEffect(() => {
+    let resizeObserver: ResizeObserver;
+    const totalTiles = typeCountTiles.length;
+
+    const updateLayout = (container: HTMLElement) => {
+      const containerWidth = container.getBoundingClientRect().width;
+      const isDesktopView = containerWidth >= minWidthDesktop;
+
+      const rowMultiplier = isDesktopView ? rowsPerView.desktop : rowsPerView.mobile;
+
+      const tilesPerRow = Math.floor(containerWidth / (tileWidth + gap));
```

Comment:
One of the tiles doesn't need gap, so this could be improved by:
`const tilesPerRow = Math.floor(containerWidth + gap / (tileWidth + gap));`

Currently `tilesPerRow` and the actual number of tiles displayed don't match sometimes, but adding "+ gap" will fix it and it will work as designed.